File: apps/frontend/src/types/external-event.ts
import type { MarketEvent } from "./market-event";

export interface ExternalEventSource {
  id: string;
  source_key: string;
  source_type: string;
  display_name: string;
  enabled: boolean;
  config?: Record<string, unknown> | null;
  last_sync_at?: string | null;
  last_error?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ExternalEventIngestionRun {
  id: string;
  source_key: string;
  status: string;
  started_at: string;
  completed_at?: string | null;
  events_found: number;
  events_created: number;
  events_skipped: number;
  error_message?: string | null;
  raw_summary?: Record<string, unknown> | null;
  created_at: string;
}

export interface ExternalEventRaw {
  id: string;
  source: string;
  external_id?: string | null;
  title: string;
  url?: string | null;
  raw_payload: Record<string, unknown>;
  normalized_market_event_id?: string | null;
  content_hash: string;
  created_at: string;
}

export interface LiveEventIngestResponse {
  run_id: string;
  source: string;
  status: string;
  events_found: number;
  events_created: number;
  events_skipped: number;
  market_events: MarketEvent[];
  run: ExternalEventIngestionRun;
  source_config?: ExternalEventSource | null;
  events: MarketEvent[];
  raw_events: ExternalEventRaw[];
  warnings: string[];
  workflows_triggered: string[];
}

export interface HackerNewsIngestRequest {
  feed: "top" | "new" | "best" | "show" | "ask" | "jobs";
  max_results: number;
  keywords?: string[] | null;
  min_score?: number | null;
  trigger_workflows?: boolean;
}

File: apps/backend/app/schemas/external_event.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.market_event import MarketEventRead


class ExternalEventSourceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    source_key: str
    source_type: str
    display_name: str
    enabled: bool
    config: dict[str, Any] | None = None
    last_sync_at: datetime | None = None
    last_error: str | None = None
    created_at: datetime
    updated_at: datetime


class ExternalEventIngestionRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    source_key: str
    status: str
    started_at: datetime
    completed_at: datetime | None = None
    events_found: int
    events_created: int
    events_skipped: int
    error_message: str | None = None
    raw_summary: dict[str, Any] | None = None
    created_at: datetime


class ExternalEventRawRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    source: str
    external_id: str | None = None
    title: str
    url: str | None = None
    raw_payload: dict[str, Any]
    normalized_market_event_id: UUID | None = None
    content_hash: str
    created_at: datetime


class LiveEventIngestRequest(BaseModel):
    query: str = "AI SaaS automation stars:>500"
    max_results: int = Field(default=10, ge=1, le=25)
    trigger_workflows: bool = False


class LiveEventIngestResponse(BaseModel):
    run_id: UUID
    source: str = "github"
    status: str
    events_found: int
    events_created: int
    events_skipped: int
    market_events: list[MarketEventRead] = Field(default_factory=list)
    run: ExternalEventIngestionRunRead
    source_config: ExternalEventSourceRead | None = None
    events: list[MarketEventRead] = Field(default_factory=list)
    raw_events: list[ExternalEventRawRead] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    workflows_triggered: list[UUID] = Field(default_factory=list)
